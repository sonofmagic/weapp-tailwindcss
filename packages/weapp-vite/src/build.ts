import type { InlineConfig } from 'vite'
import { build } from 'vite'
import type { RollupOutput, RollupWatcher } from 'rollup'
import { defu } from '@weapp-core/shared'
import { vitePluginWeapp } from './plugins'
import { getEntries } from './entry'

export function getDefaultConfig(options: { cwd?: string, entries?: string[] }): InlineConfig {
  return {
    build: {
      rollupOptions: {
        output: {
          format: 'cjs',
          entryFileNames: (chunkInfo) => {
            return chunkInfo.name
          },
        },
      },
      assetsDir: '.',
      commonjsOptions: {
        transformMixedEsModules: true,
        include: undefined,
      },
    },
    plugins: [
      vitePluginWeapp(options),
    ],
  }
}

export async function runDev(cwd: string, options?: InlineConfig) {
  const entries = await getEntries(cwd)

  if (entries) {
    const watcher = (await build(
      defu<InlineConfig, InlineConfig[]>(
        options,
        getDefaultConfig({
          cwd,
          entries: entries.all,
        }),
        {
          build: {
            watch: {},
            minify: false,
          },
        },
      )
      ,
    )) as RollupWatcher

    // watcher.on('event', (event) => {
    //   console.log(event)
    // })

    return watcher
  }
}

export async function runProd(cwd: string, options?: InlineConfig) {
  const entries = await getEntries(cwd)

  if (entries) {
    const output = (await build(
      defu<InlineConfig, InlineConfig[]>(
        options,
        getDefaultConfig({
          cwd,
          entries: entries.all,
        }),
      ),
    )) as RollupOutput | RollupOutput[]

    return output
  }
}
