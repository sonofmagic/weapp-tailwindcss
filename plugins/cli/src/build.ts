import path from 'node:path'
import defu from 'defu'
import type { FSWatcher } from 'fs-extra'
import { ensureDirSync } from 'fs-extra'
import gulp from 'gulp'
import loadPostcssConfig from 'postcss-load-config'
import type { Result } from 'postcss-load-config'
import pc from 'picocolors'
import { getPackageInfo } from 'local-pkg'
// import type { Matcher } from 'anymatch'
import { version } from '../package.json'
import { debug } from '@/debug'
import { defaultNodeModulesDirs, getDefaultOptions } from '@/defaults'
import type { BuildOptions } from '@/type'
import { getTasks } from '@/task'

export async function createBuilder(options?: Partial<BuildOptions>) {
  let postcssOptionsFromConfig: Result | undefined
  try {
    postcssOptionsFromConfig = await loadPostcssConfig({ cwd: options?.root })
  }
  catch { }
  const opt = defu<BuildOptions, Partial<BuildOptions>[]>(options, getDefaultOptions(options, postcssOptionsFromConfig))

  const { copyOthers, getCssTasks, getHtmlTasks, getJsTasks, getJsonTasks, globsSet } = await getTasks(opt)

  const tasks = {
    css: getCssTasks(),
    html: getHtmlTasks(),
    json: getJsonTasks(),
    js: getJsTasks(),
    extra: [copyOthers],
  }

  async function runTasks() {
    debug('run tasks start')
    for (const [key, value] of Object.entries(tasks)) {
      if (value) {
        debug(`run task ${pc.bold(pc.green(key))} start`)
        for (const task of value) {
          await task()
        }
        debug(`run task ${pc.bold(pc.green(key))} end`)
      }
    }
    debug('run tasks end')
  }
  const { clean, outDir, root: cwd, watchOptions } = opt
  return {
    watcher: <FSWatcher | undefined>undefined,
    async build() {
      if (clean) {
        debug('del start')
        const { deleteAsync } = await import('del')
        const patterns = [`${outDir}/**`]
        await deleteAsync(patterns, { cwd, ignore: defaultNodeModulesDirs })
        debug('del end')
      }
      ensureDirSync(path.resolve(cwd, outDir))
      await runTasks()
      return this
    },
    watch() {
      ensureDirSync(path.resolve(cwd, outDir))
      const dumps = globsSet.dump()
      const arr = (Array.isArray(watchOptions.ignored) ? watchOptions.ignored : [watchOptions.ignored]).filter(
        Boolean,
      ) as string[]
      watchOptions.ignored = [...globsSet.dumpIgnored(), ...arr]

      const watcher = gulp.watch(dumps, watchOptions, async (cb) => {
        try {
          await runTasks()
          cb()
        }
        catch (error) {
          cb(error as Error)
        }
      })
      watcher.on('change', (path) => {
        console.log(`${pc.green('changed')} ${path}`)
      })

      watcher.on('add', (path) => {
        console.log(`${pc.green('add')} ${path}`)
      })

      watcher.on('unlink', (path) => {
        console.log(`${pc.green('remove')} ${path}`)
      })

      watcher.on('ready', async () => {
        const meta = await getPackageInfo('weapp-tailwindcss')
        let weappTwVersionStr: string = ''
        if (meta) {
          weappTwVersionStr = `(${pc.blue(pc.underline(meta.version))})`
        }

        console.log(
          `${pc.bold(`${pc.green('weapp')}-${pc.blue('tailwindcss')}`)}${weappTwVersionStr} ${pc.cyan('cli')}(${pc.blue(pc.underline(version))}) is ready!`,
        )
      })
      this.watcher = watcher
      return this
    },
    globsSet,
  }
}

export async function build(options?: Partial<BuildOptions>) {
  const builder = await createBuilder(options)
  return await builder.build()
}

export async function watch(options?: Partial<BuildOptions>) {
  const builder = await createBuilder(options)
  return await builder.watch()
}
