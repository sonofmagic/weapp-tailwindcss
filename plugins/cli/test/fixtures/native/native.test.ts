import path from 'node:path'
import fs from 'fs-extra'
// import tailwindcss from 'tailwindcss'
import { build } from '@/build'

describe('build', () => {
  it('native', async () => {
    const nativePath = __dirname
    process.chdir(nativePath)
    const { globsSet } = await build({
      root: '.',
    })

    function e(...args: string[]) {
      return fs.existsSync(path.resolve(__dirname, 'dist', ...args))
    }

    function r(...args: string[]) {
      return fs.readFileSync(path.resolve(__dirname, 'dist', ...args), 'utf8')
    }

    expect(globsSet).toMatchSnapshot()
    expect(e()).toBe(true)
    expect(e('app.wxss')).toBe(true)
    expect(e('pages/index/index.js')).toBe(true)
    expect(e('pages/index/index.json')).toBe(true)
    expect(e('pages/index/index.wxml')).toBe(true)
    expect(e('pages/index/index.wxss')).toBe(true)
    expect(e('sitemap.json')).toBe(true)

    expect(r('app.wxss')).toMatchSnapshot()
    expect(r('pages/index/index.js')).toMatchSnapshot()
    expect(r('pages/index/index.json')).toMatchSnapshot()
    expect(r('pages/index/index.wxml')).toMatchSnapshot()
    expect(r('pages/index/index.wxss')).toMatchSnapshot()
    expect(r('sitemap.json')).toMatchSnapshot()
  })
})


