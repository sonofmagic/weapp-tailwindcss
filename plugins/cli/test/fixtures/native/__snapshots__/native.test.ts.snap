// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build > native 1`] = `
GlobsSet {
  "excludeSet": Set {
    "!**/node_modules/**",
    "!**/miniprogram_npm/**",
    "!**/project.config.json/**",
    "!**/project.private.config.json/**",
    "!**/package.json/**",
    "!postcss.config.js",
    "!tailwind.config.js",
    "!weapp-tw.config.js",
    "!dist/**/*",
    "!./dist/**/*",
  },
  "includeSet": Set {
    "**/*.wxss",
    "**/*.less",
    "**/*.sass",
    "**/*.scss",
    "**/*.wxml",
    "**/*.json",
    "**/*.js",
    "**/*.ts",
    "**/*.wxs",
    "**/*.{png,jpg,jpeg,gif,svg,webp}",
  },
}
`;

exports[`build > native 2`] = `
"::before,::after {
    --tw-content: ""}
view,text,::before,::after {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  ;
    --tw-contain-size:  ;
    --tw-contain-layout:  ;
    --tw-contain-paint:  ;
    --tw-contain-style:  
;
    box-sizing: border-box;
    border-width: 0;
    border-style: solid;
    border-color: currentColor}
::backdrop {
    --tw-border-spacing-x: 0;
    --tw-border-spacing-y: 0;
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x:  ;
    --tw-pan-y:  ;
    --tw-pinch-zoom:  ;
    --tw-scroll-snap-strictness: proximity;
    --tw-gradient-from-position:  ;
    --tw-gradient-via-position:  ;
    --tw-gradient-to-position:  ;
    --tw-ordinal:  ;
    --tw-slashed-zero:  ;
    --tw-numeric-figure:  ;
    --tw-numeric-spacing:  ;
    --tw-numeric-fraction:  ;
    --tw-ring-inset:  ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246 / 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur:  ;
    --tw-brightness:  ;
    --tw-contrast:  ;
    --tw-grayscale:  ;
    --tw-hue-rotate:  ;
    --tw-invert:  ;
    --tw-saturate:  ;
    --tw-sepia:  ;
    --tw-drop-shadow:  ;
    --tw-backdrop-blur:  ;
    --tw-backdrop-brightness:  ;
    --tw-backdrop-contrast:  ;
    --tw-backdrop-grayscale:  ;
    --tw-backdrop-hue-rotate:  ;
    --tw-backdrop-invert:  ;
    --tw-backdrop-opacity:  ;
    --tw-backdrop-saturate:  ;
    --tw-backdrop-sepia:  ;
    --tw-contain-size:  ;
    --tw-contain-layout:  ;
    --tw-contain-paint:  ;
    --tw-contain-style:  
}
.w-screen {
    width: 100vw
}
.grow {
    flex-grow: 1
}
.overflow-hidden {
    overflow: hidden
}
.bg-red-100 {
    --tw-bg-opacity: 1;
    background-color: rgb(254 226 226 / var(--tw-bg-opacity))
}
.bg-yellow-600 {
    --tw-bg-opacity: 1;
    background-color: rgb(202 138 4 / var(--tw-bg-opacity))
}
.p-10 {
    padding: 2.5rem
}
.pb-10 {
    padding-bottom: 2.5rem
}
.text-center {
    text-align: center
}
.text-_h123456_ {
    --tw-text-opacity: 1;
    color: rgb(18 52 86 / var(--tw-text-opacity))
}
.text-blacks90 {
    color: rgb(0 0 0 / 0.9)
}
"
`;

exports[`build > native 3`] = `
"Page({
  data: {
    motto: 'bg-red-100',
  },
})
"
`;

exports[`build > native 4`] = `
"{
  "usingComponents": {
  }
}"
`;

exports[`build > native 5`] = `
"<view class="w-screen grow overflow-hidden">
  <view class="text-blacks90">
    <view class="p-10">
      <view class="text-center">Address Book</view>
      <view class="text-_h123456_">类通讯录列表</view>
      <view class="bg-yellow-600">ICON</view>
    </view>
    <view class="pb-10 {{motto}}">motto</view>
  </view>
</view>"
`;

exports[`build > native 6`] = `""`;

exports[`build > native 7`] = `
"{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [{
  "action": "allow",
  "page": "*"
  }]
}"
`;
